NPE trying to send event without payload via CLI

Running 
`./hono-cli-2.1.0-SNAPSHOT-exec amqp-device --sandbox event`
leads to an NPE:
```
Cannot send event message.
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:413)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2118)
	at org.eclipse.hono.cli.adapter.amqp.AmqpAdapter.sendEvent(AmqpAdapter.java:609)
	at java.lang.reflect.Method.invoke(Method.java:568)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1972)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at io.quarkus.picocli.runtime.PicocliRunner$EventExecutionStrategy.execute(PicocliRunner.java:26)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at io.quarkus.picocli.runtime.PicocliRunner.run(PicocliRunner.java:40)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:129)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:208)
	at org.eclipse.hono.client.device.amqp.impl.ProtonBasedAmqpAdapterClient.sendEvent(ProtonBasedAmqpAdapterClient.java:225)
	at org.eclipse.hono.cli.adapter.amqp.AmqpAdapter.lambda$sendEvent$24(AmqpAdapter.java:595)
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.FixedMapping.onSuccess(FixedMapping.java:31)
	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
	at org.eclipse.hono.client.amqp.connection.impl.HonoConnectionImpl$ConnectionAttempt.lambda$new$0(HonoConnectionImpl.java:1004)
	at org.eclipse.hono.client.amqp.connection.impl.HonoConnectionImpl$ConnectionAttempt.lambda$connect$2(HonoConnectionImpl.java:1090)
	at io.vertx.core.impl.future.FutureImpl$1.onSuccess(FutureImpl.java:91)
	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
	at org.eclipse.hono.client.amqp.connection.impl.ConnectionFactoryImpl.lambda$handleConnectionAttemptResult$3(ConnectionFactoryImpl.java:248)
	at io.vertx.proton.impl.ProtonConnectionImpl.fireRemoteOpen(ProtonConnectionImpl.java:368)
	at io.vertx.proton.impl.ProtonTransport.handleSocketBuffer(ProtonTransport.java:116)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:98)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:239)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:129)
	at io.vertx.core.net.impl.NetSocketImpl$DataMessageHandler.handle(NetSocketImpl.java:418)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
	at io.vertx.core.impl.ContextBase.emit(ContextBase.java:239)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:394)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:833)
	at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:704)
	at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)
```

> FMPOV the latter is the correct behavior because it allows for devices to (easily) produce empty notifications ...

Yes. FMPOV, at the very least, an empty payload should be supported when using the `application/vnd.eclipse-hono-empty-notification` content type.

It even seems that the empty-notification content type is used as default when setting a payload:
```
./hono-cli-2.1.0-SNAPSHOT-exec amqp-device --sandbox event --payload=test
Cannot send event message.
java.util.concurrent.CompletionException: org.eclipse.hono.client.ClientErrorException: content type [application/vnd.eclipse-hono-empty-notification] does not match payload
```
<details>
<summary>(full stack trace)</summary>

~~~
java.util.concurrent.CompletionException: org.eclipse.hono.client.ClientErrorException: content type [application/vnd.eclipse-hono-empty-notification] does not match payload
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:413)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2118)
	at org.eclipse.hono.cli.adapter.amqp.AmqpAdapter.sendEvent(AmqpAdapter.java:609)
	at java.lang.reflect.Method.invoke(Method.java:568)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1972)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at io.quarkus.picocli.runtime.PicocliRunner$EventExecutionStrategy.execute(PicocliRunner.java:26)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at io.quarkus.picocli.runtime.PicocliRunner.run(PicocliRunner.java:40)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:129)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: org.eclipse.hono.client.ClientErrorException: content type [application/vnd.eclipse-hono-empty-notification] does not match payload
	at org.eclipse.hono.client.amqp.connection.ErrorConverter.fromTransferError(ErrorConverter.java:72)
	at org.eclipse.hono.client.amqp.connection.ErrorConverter.fromTransferError(ErrorConverter.java:50)
	at java.util.Optional.map(Optional.java:260)
	at org.eclipse.hono.client.amqp.GenericSenderLink.mapUnacceptedOutcomeToErrorResult(GenericSenderLink.java:550)
	at org.eclipse.hono.client.amqp.GenericSenderLink.lambda$sendMessageAndWaitForOutcome$10(GenericSenderLink.java:468)
	at io.vertx.proton.impl.ProtonDeliveryImpl.fireUpdate(ProtonDeliveryImpl.java:158)
	at io.vertx.proton.impl.ProtonTransport.handleSocketBuffer(ProtonTransport.java:165)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:98)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:239)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:129)
	at io.vertx.core.net.impl.NetSocketImpl$DataMessageHandler.handle(NetSocketImpl.java:418)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
	at io.vertx.core.impl.ContextBase.emit(ContextBase.java:239)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:394)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:833)
	at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:704)
	at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)
~~~

</details>

Having the empty-notification content type used as default seems odd here. I think it would be better to use `text/plain` or `application/json`, depending on the payload.

Regarding the question how to deal with an empty payload in general, there is also #3222.

--- a/cli/src/main/java/org/eclipse/hono/cli/CustomConfiguration.java
+++ b/cli/src/main/java/org/eclipse/hono/cli/CustomConfiguration.java
@@ -1,0 +1,46 @@
+/**
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+
+package org.eclipse.hono.cli;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+
+import org.eclipse.hono.cli.util.CommandUtils;
+import org.eclipse.hono.client.ServiceInvocationException;
+
+import io.quarkus.picocli.runtime.PicocliCommandLineFactory;
+import picocli.CommandLine;
+
+/**
+ * A producer for Hono specific picocli objects.
+ *
+ */
+@ApplicationScoped
+public class CustomConfiguration {
+
+    /**
+     * Creates a command line with an exception handler that understands Hono's
+     * {@link ServiceInvocationException}s.
+     *
+     * @param factory The factory to use for creating the command line.
+     * @return The command line.
+     */
+    @Produces
+    CommandLine customCommandLine(final PicocliCommandLineFactory factory) {
+        return factory.create()
+                .setCommandName("hono")
+                .setExecutionExceptionHandler(CommandUtils::handleExecutionException);
+    }
+}

--- a/cli/src/main/java/org/eclipse/hono/cli/adapter/amqp/AmqpAdapter.java
+++ b/cli/src/main/java/org/eclipse/hono/cli/adapter/amqp/AmqpAdapter.java
@@ -43,7 +43,6 @@ import org.eclipse.hono.client.amqp.connection.AmqpUtils;
 import org.eclipse.hono.client.amqp.connection.HonoConnection;
 import org.eclipse.hono.client.command.CommandConsumer;
 import org.eclipse.hono.client.device.amqp.AmqpAdapterClient;
-import org.eclipse.hono.util.EventConstants;
 import org.eclipse.hono.util.QoS;
 import org.fusesource.jansi.AnsiConsole;
 import org.jline.console.SystemRegistry;
@@ -384,15 +383,21 @@ public class AmqpAdapter implements Callable<Integer> {
         return CommandLine.ExitCode.OK;
     }
 
-    private void handleUnauthorizedError(final Throwable t, final String deviceId) {
-        if (ServiceInvocationException.extractStatusCode(t) == HttpURLConnection.HTTP_FORBIDDEN) {
-            System.err.println("""
-                    The currently connected device is not authorized to act on behalf of device [id: %1$s].
-                    In order to authorize the connected device, its device identifier needs to be added to
-                    the list of (gateway) devices that may act on behalf of device [id: %1$s].
-                    Please refer to https://www.eclipse.org/hono/docs/concepts/connecting-devices/#connecting-via-a-device-gateway
-                    for details regarding connecting devices via gateways.
-                    """.formatted(deviceId));
+    private void handleError(final Throwable t, final String deviceId) {
+        if (t instanceof ServiceInvocationException e) {
+            switch (e.getErrorCode()) {
+            case HttpURLConnection.HTTP_FORBIDDEN:
+                System.err.println("""
+                        The currently connected device is not authorized to act on behalf of device [id: %1$s].
+                        In order to authorize the connected device, its device identifier needs to be added to
+                        the list of (gateway) devices that may act on behalf of device [id: %1$s].
+                        Please refer to https://www.eclipse.org/hono/docs/concepts/connecting-devices/#connecting-via-a-device-gateway
+                        for details regarding connecting devices via gateways.
+                        """.formatted(deviceId));
+                break;
+            default:
+                System.err.println("The AMQP protocol adapter was not able to process the request.");
+            }
         }
     }
 
@@ -490,7 +495,7 @@ public class AmqpAdapter implements Callable<Integer> {
             .onSuccess(consumer -> activeConsumers.put(getConsumerKey(tenantId, deviceId), consumer))
             .onFailure(t -> {
                 System.err.println("Cannot subscribe to commands for device");
-                handleUnauthorizedError(t, deviceId);
+                handleError(t, deviceId);
             })
             .toCompletionStage()
             .toCompletableFuture()
@@ -565,15 +570,13 @@ public class AmqpAdapter implements Callable<Integer> {
             .compose(client -> client.sendTelemetry(
                     QoS.AT_MOST_ONCE,
                     Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
-                    Optional.ofNullable(options.contentType).orElseGet(() -> Optional.ofNullable(options.payload)
-                            .map(p -> (String) null)
-                            .orElse(EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION)),
+                    options.contentType,
                     options.tenantId,
                     options.deviceId,
                     null))
             .onFailure(t -> {
                 System.err.println("Cannot send telemetry message.");
-                handleUnauthorizedError(t, options.deviceId);
+                handleError(t, options.deviceId);
             })
             .toCompletionStage()
             .toCompletableFuture()
@@ -594,15 +597,13 @@ public class AmqpAdapter implements Callable<Integer> {
         getClient()
             .compose(f -> f.sendEvent(
                     Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
-                    Optional.ofNullable(options.contentType).orElseGet(() -> Optional.ofNullable(options.payload)
-                            .map(p -> (String) null)
-                            .orElse(EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION)),
+                    options.contentType,
                     options.tenantId,
                     options.deviceId,
                     null))
             .onFailure(t -> {
                 System.err.println("Cannot send event message.");
-                handleUnauthorizedError(t, options.deviceId);
+                handleError(t, options.deviceId);
             })
             .toCompletionStage()
             .toCompletableFuture()

--- a/cli/src/main/java/org/eclipse/hono/cli/app/CommandAndControl.java
+++ b/cli/src/main/java/org/eclipse/hono/cli/app/CommandAndControl.java
@@ -247,21 +247,16 @@ public class CommandAndControl implements Callable<Integer> {
         connected.set(false);
     }
 
-    private void handleCommandSendingError(final Throwable e) {
-        if (e instanceof ServiceInvocationException) {
-            final var sie = (ServiceInvocationException) e;
-            System.err.println("Error: %d %s".formatted(sie.getErrorCode(),
-                    ServiceInvocationException.getErrorMessageForExternalClient(sie)));
+    private void handleCommandSendingError(final Throwable t) {
+        System.err.println("Cannot send command to device.");
+        if (t instanceof ServiceInvocationException sie) {
             if (sie.getErrorCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
-                System.err.println("Cannot send command to device");
                 System.err.println("""
                         Check if device is connected to a protocol adapter and has subscribed \
                         for commands as described in the protocol adapter user guides.
                         https://www.eclipse.org/hono/docs/user-guide/
                         """);
             }
-        } else {
-            System.err.println("Error: failed to send command to device");
         }
     }
 
@@ -281,23 +276,19 @@ public class CommandAndControl implements Callable<Integer> {
                         .map(p -> MessageHelper.CONTENT_TYPE_APPLICATION_JSON)
                         .orElse("text/plain"));
 
-        try {
-            appCommand.getApplicationClient()
-                .compose(c -> c.sendOneWayCommand(
-                    options.tenantId,
-                    options.deviceId,
-                    options.commandName,
-                    Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
-                    ct,
-                    null))
-                .toCompletionStage()
-                .toCompletableFuture()
-                .join();
-            return ExitCode.OK;
-        } catch (final CompletionException e) {
-            handleCommandSendingError(e.getCause());
-            return ExitCode.SOFTWARE;
-        }
+        appCommand.getApplicationClient()
+            .compose(c -> c.sendOneWayCommand(
+                options.tenantId,
+                options.deviceId,
+                options.commandName,
+                Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
+                ct,
+                null))
+            .onFailure(this::handleCommandSendingError)
+            .toCompletionStage()
+            .toCompletableFuture()
+            .join();
+        return ExitCode.OK;
     }
 
     @CommandLine.Command(
@@ -322,25 +313,21 @@ public class CommandAndControl implements Callable<Integer> {
                         .map(p -> MessageHelper.CONTENT_TYPE_APPLICATION_JSON)
                         .orElse("text/plain"));
 
-        try {
-            appCommand.getApplicationClient()
-                .compose(c -> c.sendCommand(
-                    options.tenantId,
-                    options.deviceId,
-                    options.commandName,
-                    Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
-                    ct,
-                    UUID.randomUUID().toString(),
-                    Duration.ofSeconds(responseTimeout),
-                    null))
-                .onSuccess(this::printResponse)
-                .toCompletionStage()
-                .toCompletableFuture()
-                .join();
-            return ExitCode.OK;
-        } catch (final CompletionException e) {
-            handleCommandSendingError(e.getCause());
-            return ExitCode.SOFTWARE;
-        }
+        appCommand.getApplicationClient()
+            .compose(c -> c.sendCommand(
+                options.tenantId,
+                options.deviceId,
+                options.commandName,
+                Optional.ofNullable(options.payload).map(Buffer::buffer).orElse(null),
+                ct,
+                UUID.randomUUID().toString(),
+                Duration.ofSeconds(responseTimeout),
+                null))
+            .onSuccess(this::printResponse)
+            .onFailure(this::handleCommandSendingError)
+            .toCompletionStage()
+            .toCompletableFuture()
+            .join();
+        return ExitCode.OK;
     }
 }

--- a/cli/src/main/java/org/eclipse/hono/cli/util/CommandUtils.java
+++ b/cli/src/main/java/org/eclipse/hono/cli/util/CommandUtils.java
@@ -46,8 +46,9 @@ public final class CommandUtils {
      */
     public static void printError(final Throwable t) {
         if (t instanceof ServiceInvocationException cause) {
-            System.err.printf("Error: %d %s%n", cause.getErrorCode(),
-                    ServiceInvocationException.getErrorMessageForExternalClient(cause));
+            System.err.println("Error: %d - %s".formatted(
+                    cause.getErrorCode(),
+                    ServiceInvocationException.getErrorMessageForExternalClient(cause)));
         } else {
             System.err.println("Error: %s".formatted(t.getMessage()));
         }

--- a/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/EventSender.java
+++ b/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/EventSender.java
@@ -30,7 +30,7 @@ public interface EventSender {
      *
      * @param payload The data to send.
      *            <p>
-     *            The payload will be added to the message as an AMQP 1.0 <em>Data</em> section.
+     *            The payload, if not {@code null}, will be added to the message's body as an AMQP 1.0 <em>Data</em> section.
      * @param contentType The content type of the payload or {@code null} if unknown.
      *            <p>
      *            This parameter will be used as the value for the message's <em>content-type</em> property.
@@ -54,7 +54,6 @@ public interface EventSender {
      *         the future will be failed with either a {@code org.eclipse.hono.client.ServerErrorException} or a
      *         {@link org.eclipse.hono.client.ClientErrorException} depending on the reason for the failure to
      *         process the message.
-     * @throws NullPointerException if payload is {@code null}.
      * @throws IllegalArgumentException if tenant ID is not {@code null} but device ID is {@code null}.
      */
     Future<ProtonDelivery> sendEvent(

--- a/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/TelemetrySender.java
+++ b/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/TelemetrySender.java
@@ -33,7 +33,7 @@ public interface TelemetrySender {
      * @param qos The delivery semantics to use.
      * @param payload The data to send.
      *            <p>
-     *            The payload will be added to the message as an AMQP 1.0 <em>Data</em> section.
+     *            The payload, if not {@code null}, will be added to the message as an AMQP 1.0 <em>Data</em> section.
      * @param contentType The content type of the payload or {@code null} if unknown.
      *            <p>
      *            This parameter will be used as the value for the message's <em>content-type</em> property.
@@ -58,7 +58,7 @@ public interface TelemetrySender {
      *         will be failed with either a {@code org.eclipse.hono.client.ServerErrorException} or a
      *         {@link org.eclipse.hono.client.ClientErrorException} depending on the reason for the
      *         failure to process the message.
-     * @throws NullPointerException if quality-of-service or payload are {@code null}.
+     * @throws NullPointerException if quality-of-service is {@code null}.
      * @throws IllegalArgumentException if tenant ID is not {@code null} but device ID is {@code null}.
      */
     Future<ProtonDelivery> sendTelemetry(

--- a/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/impl/ProtonBasedAmqpAdapterClient.java
+++ b/clients/device-amqp/src/main/java/org/eclipse/hono/client/device/amqp/impl/ProtonBasedAmqpAdapterClient.java
@@ -14,8 +14,11 @@
 package org.eclipse.hono.client.device.amqp.impl;
 
 import java.util.Objects;
+import java.util.Optional;
 import java.util.function.Consumer;
 
+import org.apache.qpid.proton.amqp.Binary;
+import org.apache.qpid.proton.amqp.messaging.Data;
 import org.apache.qpid.proton.message.Message;
 import org.eclipse.hono.client.amqp.GenericSenderLink;
 import org.eclipse.hono.client.amqp.connection.AmqpUtils;
@@ -172,7 +175,12 @@ public final class ProtonBasedAmqpAdapterClient extends ConnectionLifecycleWrapp
         final Message msg = ProtonHelper.message();
         msg.setAddress(targetAddress);
         AmqpUtils.setCreationTime(msg);
-        AmqpUtils.setPayload(msg, contentType, payload);
+        Optional.ofNullable(contentType).ifPresent(msg::setContentType);
+        Optional.ofNullable(payload)
+            .map(Buffer::getBytes)
+            .map(Binary::new)
+            .map(Data::new)
+            .ifPresent(msg::setBody);
         return msg;
     }
 
@@ -192,7 +200,6 @@ public final class ProtonBasedAmqpAdapterClient extends ConnectionLifecycleWrapp
             final SpanContext context) {
 
         Objects.requireNonNull(qos);
-        Objects.requireNonNull(payload);
 
         checkDeviceSpec(tenantId, deviceId);
         return getOrCreateGenericTelemetrySender()
@@ -221,8 +228,6 @@ public final class ProtonBasedAmqpAdapterClient extends ConnectionLifecycleWrapp
             final String tenantId,
             final String deviceId,
             final SpanContext context) {
-
-        Objects.requireNonNull(payload);
 
         checkDeviceSpec(tenantId, deviceId);
         return getOrCreateGenericEventSender()

--- a/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/AmqpAdapterClientTestBase.java
+++ b/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/AmqpAdapterClientTestBase.java
@@ -50,8 +50,6 @@ public abstract class AmqpAdapterClientTestBase {
 
     protected static final String TENANT_ID = "test-tenant";
     protected static final String DEVICE_ID = "test-device";
-    protected static final String CONTENT_TYPE = "text/plain";
-    protected static final Buffer PAYLOAD = Buffer.buffer("test-value");
     protected static final String TEST_PROPERTY_KEY = "test-key";
     protected static final String TEST_PROPERTY_VALUE = "test-value";
 
@@ -109,10 +107,15 @@ public abstract class AmqpAdapterClientTestBase {
      * Executes the assertions that check that the message created by the client conforms to the expectations of the
      * AMQP adapter.
      *
-     * @param expectedAddress The expected target address.
+     * @param expectedAddress The target address expected to be found in the message that has been sent.
+     * @param expectedContentType The content type expected to be found in the message that has been sent.
+     * @param expectedPayload The payload expected to be found in the message that has been sent.
      * @return The captured message.
      */
-    protected Message assertMessageConformsAmqpAdapterSpec(final String expectedAddress) {
+    protected Message assertMessageConformsToAmqpAdapterSpec(
+            final String expectedAddress,
+            final String expectedContentType,
+            final Buffer expectedPayload) {
 
         final ArgumentCaptor<Message> messageArgumentCaptor = ArgumentCaptor.forClass(Message.class);
 
@@ -120,8 +123,20 @@ public abstract class AmqpAdapterClientTestBase {
         final Message message = messageArgumentCaptor.getValue();
 
         assertAll(() -> assertThat(message.getAddress()).isEqualTo(expectedAddress),
-                () -> assertThat(AmqpUtils.getPayload(message)).isEqualTo(PAYLOAD),
-                () -> assertThat(message.getContentType()).isEqualTo(CONTENT_TYPE));
+                () -> {
+                    if (expectedContentType != null) {
+                        assertThat(message.getContentType()).isEqualTo(expectedContentType);
+                    } else {
+                        assertThat(message.getContentType()).isNull();
+                    }
+                },
+                () -> {
+                    if (expectedPayload != null) {
+                        assertThat(AmqpUtils.getPayload(message)).isEqualTo(expectedPayload);
+                    } else {
+                        assertThat(AmqpUtils.getPayload(message)).isNull();
+                    }
+                });
 
         return message;
     }

--- a/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/impl/CommandResponderTest.java
+++ b/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/impl/CommandResponderTest.java
@@ -21,16 +21,17 @@ import static org.mockito.Mockito.verify;
 
 import static com.google.common.truth.Truth.assertThat;
 
-import org.apache.qpid.proton.message.Message;
+import java.util.Optional;
+
 import org.eclipse.hono.client.device.amqp.AmqpAdapterClientTestBase;
-import org.eclipse.hono.client.device.amqp.impl.ProtonBasedAmqpAdapterClient;
 import org.eclipse.hono.util.CommandConstants;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.ValueSource;
+import org.junit.jupiter.params.provider.CsvSource;
 
 import io.opentracing.SpanContext;
+import io.vertx.core.buffer.Buffer;
 import io.vertx.junit5.VertxExtension;
 import io.vertx.junit5.VertxTestContext;
 
@@ -58,32 +59,36 @@ public class CommandResponderTest extends AmqpAdapterClientTestBase {
         client = new ProtonBasedAmqpAdapterClient(connection);
     }
 
-    private void assertMessageConformsAmqpAdapterSpec() {
-        final Message message = assertMessageConformsAmqpAdapterSpec(ADDRESS.toString());
-        assertThat(message.getCorrelationId()).isEqualTo(CORRELATION_ID);
-        assertThat(message.getApplicationProperties().getValue().get("status")).isEqualTo(STATUS);
-    }
-
     /**
      * Verifies that a command response message sent by the client conforms to the expectations of the AMQP adapter.
      *
+     * @param payload The payload to put in the message body.
+     * @param contentType The value to set as the message's content-type.
      * @param useSpanContext {@code true} if the sending should be tracked.
      * @param ctx The test context to use for running asynchronous tests.
      */
     @ParameterizedTest
-    @ValueSource(booleans = { true, false })
+    @CsvSource(value = {
+        "the-payload,custom/content,true",
+        "the-payload,,false",
+        ",custom/content,true",
+        ",,false",
+    })
     public void testSendCommandResponseCreatesValidMessage(
+            final String payload,
+            final String contentType,
             final boolean useSpanContext,
             final VertxTestContext ctx) {
 
         final var spanContext = mock(SpanContext.class);
+        final var expectedBody = Optional.ofNullable(payload).map(Buffer::buffer).orElse(null);
         // WHEN sending a command response using the API
         final var result = client.sendCommandResponse(
                 ADDRESS,
                 CORRELATION_ID,
                 STATUS,
-                PAYLOAD,
-                CONTENT_TYPE,
+                expectedBody,
+                contentType,
                 useSpanContext ? spanContext : null);
 
         // THEN the future waits for the disposition to be updated by the peer
@@ -94,7 +99,9 @@ public class CommandResponderTest extends AmqpAdapterClientTestBase {
         result.onComplete(ctx.succeeding(ok -> {
                 // THEN the AMQP message conforms to the expectations of the AMQP protocol adapter
                 ctx.verify(() -> {
-                    assertMessageConformsAmqpAdapterSpec();
+                    final var message = assertMessageConformsToAmqpAdapterSpec(ADDRESS.toString(), contentType, expectedBody);
+                    assertThat(message.getCorrelationId()).isEqualTo(CORRELATION_ID);
+                    assertThat(message.getApplicationProperties().getValue().get("status")).isEqualTo(STATUS);
                     if (useSpanContext) {
                         // and the given SpanContext is used
                         verify(spanBuilder).addReference(any(), eq(spanContext));

--- a/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/impl/TelemetrySenderTest.java
+++ b/clients/device-amqp/src/test/java/org/eclipse/hono/client/device/amqp/impl/TelemetrySenderTest.java
@@ -21,7 +21,8 @@ import static org.mockito.Mockito.verify;
 
 import static com.google.common.truth.Truth.assertThat;
 
-import org.apache.qpid.proton.message.Message;
+import java.util.Optional;
+
 import org.eclipse.hono.client.device.amqp.AmqpAdapterClientTestBase;
 import org.eclipse.hono.util.EventConstants;
 import org.eclipse.hono.util.QoS;
@@ -57,27 +58,11 @@ public class TelemetrySenderTest extends AmqpAdapterClientTestBase {
     }
 
     /**
-     * Executes the assertions that check that the message created by the client conforms to the expectations of the
-     * AMQP adapter.
-     *
-     * @param endpoint The expected target address endpoint.
-     * @param tenantId The expected target address tenant.
-     * @param deviceId The expected target address device.
-     * @return The captured message.
-     */
-    private Message assertMessageConformsAmqpAdapterSpec(
-            final String endpoint,
-            final String tenantId,
-            final String deviceId) {
-
-        final var expectedAddress = ResourceIdentifier.fromPath(endpoint, tenantId, deviceId).toString();
-        return assertMessageConformsAmqpAdapterSpec(expectedAddress);
-    }
-
-    /**
      * Verifies that a telemetry message sent by the client conforms to the expectations of the AMQP adapter.
      *
      * @param qos The delivery semantics.
+     * @param payload The payload to put in the message body.
+     * @param contentType The value to set as the message's content-type.
      * @param tenantId The tenant that the device belongs to.
      * @param deviceId The identifier of the device.
      * @param useSpanContext {@code true} if the sending should be tracked.
@@ -85,22 +70,28 @@ public class TelemetrySenderTest extends AmqpAdapterClientTestBase {
      */
     @ParameterizedTest
     @CsvSource(value = {
-                            "AT_MOST_ONCE,,my-device,true", "AT_MOST_ONCE,my-tenant,my-device,false",
-                            "AT_LEAST_ONCE,,my-device,false", "AT_LEAST_ONCE,my-tenant,my-device,true",
+        "AT_MOST_ONCE,the-payload,,,my-device,true",
+        "AT_MOST_ONCE,,custom/content,my-tenant,my-device,false",
+        "AT_LEAST_ONCE,the-payload,custom/content,,my-device,false",
+        "AT_LEAST_ONCE,,,my-tenant,my-device,true",
     })
     public void testSendTelemetryCreatesValidMessage(
             final QoS qos,
+            final String payload,
+            final String contentType,
             final String tenantId,
             final String deviceId,
             final boolean useSpanContext,
             final VertxTestContext ctx) {
 
         final var spanContext = mock(SpanContext.class);
+        final var expectedBody = Optional.ofNullable(payload).map(Buffer::buffer).orElse(null);
+
         // WHEN sending a message using the API
         final var result = client.sendTelemetry(
                 qos,
-                PAYLOAD,
-                CONTENT_TYPE,
+                expectedBody,
+                contentType,
                 tenantId,
                 deviceId,
                 useSpanContext ? spanContext : null);
@@ -115,10 +106,10 @@ public class TelemetrySenderTest extends AmqpAdapterClientTestBase {
         result.onComplete(ctx.succeeding(ok -> {
                 // THEN the AMQP message conforms to the expectations of the AMQP protocol adapter
                 ctx.verify(() -> {
-                    assertMessageConformsAmqpAdapterSpec(
-                            TelemetryConstants.TELEMETRY_ENDPOINT,
-                            tenantId,
-                            deviceId);
+                    assertMessageConformsToAmqpAdapterSpec(
+                            ResourceIdentifier.fromPath(TelemetryConstants.TELEMETRY_ENDPOINT, tenantId, deviceId).toString(),
+                            contentType,
+                            expectedBody);
                     if (useSpanContext) {
                         // and the given SpanContext is used
                         verify(spanBuilder).addReference(any(), eq(spanContext));
@@ -139,24 +130,33 @@ public class TelemetrySenderTest extends AmqpAdapterClientTestBase {
     /**
      * Verifies that an event message sent by the client conforms to the expectations of the AMQP adapter.
      *
+     * @param payload The payload to put in the message body.
+     * @param contentType The value to set as the message's content-type.
      * @param tenantId The tenant that the device belongs to.
      * @param deviceId The identifier of the device.
      * @param useSpanContext {@code true} if the sending should be tracked.
      * @param ctx The test context to use for running asynchronous tests.
      */
     @ParameterizedTest
-    @CsvSource(value = { ",,true", ",other-device,false", "my-tenant,my-device,true" })
+    @CsvSource(value = {
+        "the-payload,custom/content,,,true",
+        ",,,other-device,false",
+        "the-payload,,my-tenant,my-device,true"
+    })
     public void testSendEventCreatesValidMessage(
+            final String payload,
+            final String contentType,
             final String tenantId,
             final String deviceId,
             final boolean useSpanContext,
             final VertxTestContext ctx) {
 
         final var spanContext = mock(SpanContext.class);
+        final var expectedBody = Optional.ofNullable(payload).map(Buffer::buffer).orElse(null);
         // WHEN sending a message using the API
         final var result = client.sendEvent(
-                PAYLOAD,
-                CONTENT_TYPE,
+                expectedBody,
+                contentType,
                 tenantId,
                 deviceId,
                 useSpanContext ? spanContext : null);
@@ -169,10 +169,10 @@ public class TelemetrySenderTest extends AmqpAdapterClientTestBase {
         result.onComplete(ctx.succeeding(ok -> {
                 // THEN the AMQP message conforms to the expectations of the AMQP protocol adapter
                 ctx.verify(() -> {
-                    assertMessageConformsAmqpAdapterSpec(
-                            EventConstants.EVENT_ENDPOINT,
-                            tenantId,
-                            deviceId);
+                    assertMessageConformsToAmqpAdapterSpec(
+                            ResourceIdentifier.fromPath(EventConstants.EVENT_ENDPOINT, tenantId, deviceId).toString(),
+                            contentType,
+                            expectedBody);
                     if (useSpanContext) {
                         // and the given SpanContext is used
                         verify(spanBuilder).addReference(any(), eq(spanContext));

